LilyPad: Add separate bindings

FlatOutPS2 opened the pull request:
**LilyPad: Add separate bindings for each pad type**

Adds separate bindings for each of the pad types (DualShock2, Guitar,
Pop'n Music). This allows the user to change the button configuration to
better suit the Guitar and Pop'n Music pads without messing up the
bindings already setup for the DS2.

Fullfills request: #1576

FlatOutPS2 commented:
What do to with the Linux side? It fails the build because it uses some of the same files, which are now incompatible.

willkuer commented:
Virtual machine is your friends...

FlatOutPS2 commented:
I don't have the the disc space for a Linux build enviroment. It would be easier to move old versions of the modified LilyPad files to the Linux folder.

gregory38 commented:
sorry for the extra work. maybe one day someone will finish the job.

MrCK1 commented:
Separate bindings work great. Nice job!

FlatOutPS2 commented:
> Edit: although bindings are separated properly now, the indicator showing the # of bindings is still constant between both modes. I don't know if we want to do anything with that?

Do you mean the # of bindings in the pad selection area of the Pads section on the General tab? Because that works perfectly for me.

MrCK1 commented:
@FlatOutPS2 Yes I did. Never mind I guess 😛

FlatOutPS2 commented:
@MrCK1 So was this a mistake or does it still not behave correctly?

MrCK1 commented:
What I meant was that Lilypad could say that there are 39 bindings, but number is the same in Dualshock 2 and Guitar Since the bindings are separated now by mode, the number of displayed bindings should change with the selected mode.

FlatOutPS2 commented:
Yes, I know. And like I said it works on my side. But if it doesn't work for you, we need to look into it. Can you share your LilyPad.ini file and can you try if it works correctly with a fresh ini file?

MrCK1 commented:
Deleting the .ini didn't work. I also noticed that any deleted binding is removed in all modes

[Here's the .ini](http://pastebin.com/gPmcQJ0H) - I also left the deleted binding (R1) removed

FlatOutPS2 commented:
I can't reproduce your issue, with or without your ini. I even tried it on a different PC and it worked without issue too. 

Is that the ini before you used it with my PR? It only contains one profile for pad 1 DualShock2 in the old format that hasn't been saved by the LilyPad from this PR.

FlatOutPS2 commented:
The Travis-ci linux build seem to fail because a few files aren't in clang format, but it doesn't specify anything. :/

gregory38 commented:
I will improve the logging. You can search `did not match`  in the build log
```
file plugins/LilyPad/Config.cpp did not match clang-format
file plugins/LilyPad/DirectInput.cpp did not match clang-format
file plugins/LilyPad/Linux/Config.cpp did not match clang-format
file plugins/LilyPad/LilyPad.cpp did not match clang-format
```
You need to install clang 3.8 plugin to MSVC. And then use the default PCSX2 formatting.
http://llvm.org/builds/
```
Go to Tools->Options->LLVM/Clang->ClangFormat and put file in the Style option field. (could be optional)
```

gregory38 commented:
@turtleli do you have a link to the version 3.8 of clang-format for windows. My above link is latest svn.

FlatOutPS2 commented:
@gregory38 Thanks. I found the 3.8 version here: http://llvm.org/releases/download.html

I did the see the files mentioned, but what I meant was it didn't specify what's wrong inside those files. It should only be a few minor issues.

gregory38 commented:
Ok. I would change the script to give more details (hopefully it isn't a clang/linux/msvc version issue)

FlatOutPS2 commented:
> Edit: did you try to run clang-format on those files ?

I tried, but nothings happens when I apply it. :/

gregory38 commented:
I was afraid of this behavior. Which version did you take ? 3.8 or 3.8.1 ?

FlatOutPS2 commented:
Yeah, turns out I selected a slightly newer version of the plugin by mistake.

gregory38 commented:
Which one exactly? I need to check your latest commit too. Maybe we need to pack the plugin in git. I need to think about it.

FlatOutPS2 commented:
I'd have to look it. The VS plugin wasn't included in the 3.8.0 build, so I took the plugin version from the same date as the 3.8.0 release. The first time I downloaded a version from the wrong date (slightly newer).

gregory38 commented:
It seems like 3.8 got some bugs that were fixed a bit later. However the ubuntu/debian version is 3.8.0. I need to check the behavior of 3.9.

gregory38 commented:
I try 3.9. But behavior is still bad...
The only different is this one
```
diff --git a/plugins/LilyPad/WndProcEater.cpp b/plugins/LilyPad/WndProcEater.cpp
index 0379005..a4a9f5d 100644
--- a/plugins/LilyPad/WndProcEater.cpp
+++ b/plugins/LilyPad/WndProcEater.cpp
@@ -127,7 +127,7 @@ bool WndProcEater::SetWndHandle(HWND hWnd)
     //printf( "(Lilypad) (Re)-Setting window handle! -> this=0x%08x, hWnd=0x%08x\n", this, hWnd );
 
     Release();
-    SetProp(hWnd, L"LilyHaxxor", (HANDLE) this);
+    SetProp(hWnd, L"LilyHaxxor", (HANDLE)this);
 
     eatenWndProc = (WNDPROC)SetWindowLongPtr(hWnd, GWLP_WNDPROC, (LONG_PTR)OverrideWndProc);
     hWndEaten = (eatenWndProc) ? hWnd : 0;
```
I didn't review whitespace and tabulation. I guess I need to create a script to check the behavior.

gregory38 commented:
@FlatOutPS2 
Could you create a github wiki page with the link to the clang-format plugin. And how to install/configure it on windows. So I could forward it to @ramapcsx2

FlatOutPS2 commented:
> Could you create a github wiki page with the link to the clang-format plugin. And how to install/configure it on windows. So I could forward it to @ramapcsx2 

I would, but the problem is I have't found a stable plugin. The older plugin I found worked for LilyPad, but might still differ from 3.8 when applying it to other parts of the PCSX2 code.

FlatOutPS2 commented:
@gregory38 Added it here: https://github.com/PCSX2/pcsx2/wiki/Clang-Format-Visual-Studio-plug-in

FlatOutPS2 commented:
I've decided to add the other features as a separate PR after this has been merged.

turtleli commented on plugins/LilyPad/Config.cpp:
```diff
@@ -933,9 +945,8 @@ int LoadSettings(int force, wchar_t *file)
-    for (int i = 0; i < sizeof(BoolOptionsInfo) / sizeof(BoolOptionsInfo[0]); i++) {
+    for (int i = 0; i < sizeof(BoolOptionsInfo) / sizeof(BoolOptionsInfo[0]); i++)
         config.bools[i] = GetPrivateProfileBool(L"General Settings", BoolOptionsInfo[i].name, BoolOptionsInfo[i].defaultValue, file);
-    }
```
While I personally prefer not using braces for single statement if/else/for/while, this is kind of a pointless change since you're not changing anything else and should be avoided in large commits.

turtleli commented on plugins/LilyPad/Config.cpp:
```diff
@@ -946,9 +957,8 @@ int LoadSettings(int force, wchar_t *file)
     config.mouseApi = (DeviceAPI)GetPrivateProfileIntW(L"General Settings", L"Mouse Mode", 0, file);
 
-    if (config.debug) {
+    if (config.debug)
```
Same here.

turtleli commented on plugins/LilyPad/Linux/Config.cpp:
```diff
@@ -374,21 +382,30 @@ int LoadSettings(int force, wchar_t *file)
             unsigned int uid;
-            int port, command, sensitivity, turbo, slot = 0, deadZone = 0;
+            int port, command, sensitivity, turbo, slot = 0, deadZone = 0, padtype = 0;
+            ;
```
Hmm.

FlatOutPS2 replied:
Hmm indeed. Don't know how that got there.

turtleli commented on plugins/LilyPad/LilyPad.cpp:
```diff
@@ -815,9 +817,8 @@ int saveStateIndex = 0;
-    if (LoadSettings() < 0) {
+    if (LoadSettings() < 0)
         return -1;
-    }
```
Unnecessary changes here and below too.

FlatOutPS2 commented on plugins/LilyPad/Linux/Config.cpp:
```diff
@@ -374,21 +382,30 @@ int LoadSettings(int force, wchar_t *file)
                     c = dev->AddVirtualControl(uid, -1);
                 if (c) {
-                    BindCommand(dev, uid, port, slot, command, sensitivity, turbo, deadZone);
+                    if (padtype == NULL) { // If ini file is imported from older version, make sure bindings aren't applied to "Unplugged" padtype.
```
Good point. :p

FlatOutPS2 commented on plugins/LilyPad/LilyPad.cpp:
```diff
@@ -511,7 +512,7 @@ void Update(unsigned int port, unsigned int slot)
     LastCheck = t;
 
-    int i;
+    int i, padtype;
```
I moved some stuff around. Must've missed this one.

FlatOutPS2 commented on plugins/LilyPad/Config.h:
```diff
@@ -21,13 +21,13 @@
 extern u8 ps2e;
 
 enum PadType {
+    // Changing the number of PadTypes requires numPadTypes in InputManager.h to be modified as well.
```
And how would you count it?

ssakash replied:
Maybe turtleli meant something like this?

Currently
> DisabledPad = 0,
     Dualshock2Pad = 1,
     GuitarPad = 2,
     PopnPad = 3

If you add a new enum called ``Pad Count`` then,

>> DisabledPad = 0,
     Dualshock2Pad = 1,
     GuitarPad = 2,
     PopnPad = 3,
     PadCount = 4 // Counts the total number of pads and is flexible as it adjusts itself when more enums are added

FlatOutPS2 replied:
Yes, I thought of that as well. But the use of fake enums is usually frowned upon. :p

turtleli replied:
Well, if you can think of a better way, go for it? ;)

turtleli commented on plugins/LilyPad/XInputEnum.cpp:
```diff
@@ -113,31 +114,31 @@ class XInputDevice : public Device
+                L"Left Thumb Y",
+                L"Right Thumb X",
+                L"Right Thumb Y",
+            };
```
Still don't know why this was moved. IMO it would actually be better left where it originally was and `i < 21` changed into `i < (sizeof(names) / sizeof(names[0]))` or similar (I didn't check if it works).

turtleli commented on plugins/LilyPad/DirectInput.cpp:
```diff
@@ -220,7 +222,7 @@ class DirectInputDevice : public Device
             format.dwNumObjs = numPhysicalControls;
             format.rgodf = formats;
-            int res = did->SetDataFormat(&format);
+            // int res = did->SetDataFormat(&format); // is not used
```
Are you sure? `res` doesn't look like it's used anywhere, but `did->SetDataFormat()` seems important (I don't know DirectInput stuff). If this line is commented out, then setting the members of `format` then becomes useless.

FlatOutPS2 replied:
Yeah, `did->SetDataFormat()` is required. I should've only removed the `int rest = `.

turtleli commented on plugins/LilyPad/DualShock3.cpp:
```diff
@@ -208,13 +208,13 @@ class DualShock3Device : public Device
     OVERLAPPED writeop;
     int writeCount;
-    int lastWrite;
+    int lastWrite = NULL;
```
IMO it's more readable if you use 0 instead of NULL when the type is not a pointer.

At the very least, I associate `NULL` with pointers.

turtleli commented on plugins/LilyPad/Linux/Config.cpp:
```diff
@@ -81,29 +81,31 @@ struct GeneralSettingsBool
-    memmove(bindings + i, bindings + i + 1, sizeof(Binding) * (dev->pads[port][slot].numBindings - i - 1));
-    dev->pads[port][slot].numBindings--;
+    if (padtype == active_padtype) // Only do the memmove when it is the active padtype
+        memmove(bindings + i, bindings + i + 1, sizeof(Binding) * (dev->pads[port][slot][padtype].numBindings - i - 1));
```
This still needs fixed.

Apart from that it seems fine and I see no obvious errors (I haven't tested the code myself though).

FlatOutPS2 commented:
I've made a change to Config.cpp to prevent to warn the user of conflicting/deleted bindings when swapping profiles.

turtleli commented:
Would it be better if the swap bindings code swapped all bindings on one pad with the respective bindings on the other pad (DS2 with DS2, guitar with guitar, and so on)?

I'm not sure what your current swap bindings code is supposed to do (Lilypad code isn't too readable for me, hence why I take donkeys to review the PR).
Test case:
1. Start with no Lilypad.ini
2. Open Lilypad config - Both pads set to DS2 and have 27 bindings
3. Swap bindings, confirm change -> Now both pads have 11 bindings -> Makes no sense to me.

FlatOutPS2 commented:
> Swap bindings, confirm change -> Now both pads have 11 bindings -> Makes no sense to me.

I changed part of the code last minute to streamline it and that caused the issue. It should normally only ever delete from one of the pads.

I've changed the behaviour. Swapping bindings will no longer let you delete any bindings. 
It now works like this:
Default behaviour: Swap pad1 with pad2 swaps all bindings from all pad types for both pads (like you said).
New "Swap Inidivdual pad types" option allows swapping and clearing inidivual pad types.

Example to get delete warning (which should rarely happen):
1. Both pads set to DS2, change pad1 pad type to Guitar.
2. Code checks if pad1 guitar has conflicting bindings with pad2 DS2 (if multiple bindings option is disabled).
3. If so , show a warning message explaining the issue and ask if you want to delete the conflicting binding from the newly selected pad type.
4. You choose :
- Yes delete the conflicting bindings and switch to new pad type.
- No do not delete anything and switch to new padtype.
- Cancel and return to the previously selected padtype without doing anything.

turtleli commented:
I tested before you made your latest edits and it seemed to work ok.

Would it be better to have both "all swap/clear" and "individual swap/clear" in the pop up menu at the same time instead of having to check/uncheck the checkbox to change the behaviour?

P.S. If you make further changes, do it in a separate commit. It means I have less to review and we can get this merged quicker. I can always squash merge on my end.

turtleli commented on plugins/LilyPad/Config.cpp:
```diff
@@ -2079,26 +2200,50 @@ INT_PTR CALLBACK GeneralDialogProc(HWND hWnd, unsigned int msg, WPARAM wParam, L
-                            PadBindings bindings = dm->devices[i]->pads[port1][slot1];
-                            dm->devices[i]->pads[port1][slot1] = dm->devices[i]->pads[port2][slot2];
-                            dm->devices[i]->pads[port2][slot2] = bindings;
+                            if (config.swapPadtypes) {
```
`swapPadTypes` isn't great naming if it's changing between individual binding swap and all bindings swap IMO.

FlatOutPS2 commented:
> Would it be better to have both "all swap/clear" and "individual swap/clear" in the pop up menu at the same time instead of having to check/uncheck the checkbox to change the behaviour?

Yeah, I had some issues with it before, but I've managed to make it work.

turtleli commented:
Something's not quite right.

Test case (starting with default settings, both DS2):
1. Click port 1 multitap.
2. Right click PAD 1B, choose "Swap with Pad 1A, DualShock 2 bindings only" => Pad 1B DS2 bindings swap with Pad 2 DS2 bindings.

If you want, you can always keep just the "Swap all bindings options" for this PR and then add individual swap in a later PR (entirely up to you). I think some of the other LilyPad changes you wanted to make probably could have been done in parallel if this PR wasn't holding it up.

FlatOutPS2 commented:
> If you want, you can always keep just the "Swap all bindings options" for this PR and then add individual swap in a later PR (entirely up to you). I think some of the other LilyPad changes you wanted to make probably could have been done in parallel if this PR wasn't holding it up.

I fixed the issue you described. It was a very small placement error that only negatively affected this one option. :p I'd rather have everything that's now in the PR function correctly, then postponing it. The other changes will come later as I will make a tooltips PR next.

FlatOutPS2 commented:
> If you want, you can always keep just the "Swap all bindings options" for this PR and then add individual swap in a later PR (entirely up to you). I think some of the other LilyPad changes you wanted to make probably could have been done in parallel if this PR wasn't holding it up.

I fixed the issue you described. It was a very small placement error that only negatively affected this one option. :p I'd prefer to have everything that's now in the PR functioning correctly, rather then postponing it. The other changes will come later as I will make a tooltips PR next.

turtleli commented on plugins/LilyPad/Linux/Config.cpp:
```diff
@@ -167,20 +167,22 @@ int BindCommand(Device *dev, unsigned int uid, unsigned int port, unsigned int s
-                        i--;
+                for (int padtype2 = 0; padtype2 < numPadTypes; padtype2++) {
+                    if (port2 == (int)port && slot2 == (int)slot)
+                        continue;
```
`if` should be moved outside the loop.

turtleli commented on plugins/LilyPad/Default.ini:
```diff
@@ -10,6 +10,7 @@ XInput=1
 DualShock 3=0
 Multitap 1=0
 Multitap 2=0
+Swap Pad Types=0
```
Needs removed.
